src/hotspot/share/gc/g1/g1ConcurrentRefine.hpp

```diff
@@ -26,132 +26,194 @@
+  // precondition: current thread is not the designated worker.
+  bool activate(uint worker_id);
 
   void worker_threads_do(ThreadClosure* tc);
   void stop();
 };
 
-// Controls refinement threads and their activation based on the number of
-// cards currently available in the global dirty card queue.
-// Refinement threads obtain work from the queue (a buffer at a time) based
-// on these thresholds. They are activated gradually based on the amount of
-// work to do.
-// Refinement thread n activates thread n+1 if the instance of this class determines there
-// is enough work available. Threads deactivate themselves if the current amount of
-// available cards falls below their individual threshold.
+// Controls concurrent refinement.
+//
+// Mutator threads produce dirty cards, which need to be examined for updates
+// to the remembered sets (refinement).  There is a pause-time budget for
+// processing these dirty cards (see -XX:G1RSetUpdatingPauseTimePercent).  The
+// purpose of concurrent refinement is to (attempt to) ensure the number of
+// pending dirty cards at the start of a GC can be processed within that time
+// budget.
+//
+// Concurrent refinement is performed by a combination of dedicated threads
+// and by mutator threads as they produce dirty cards.  If configured to not
+// have any dedicated threads (-XX:G1ConcRefinementThreads=0) then all
+// concurrent refinement work is performed by mutator threads.  When there are
+// dedicated threads, they generally do most of the concurrent refinement
+// work, to minimize throughput impact of refinement work on mutator threads.
+//
+// This class determines the target number of dirty cards pending for the next
+// GC.  It also owns the dedicated refinement threads and controls their
+// activation in order to achieve that target.
+//
+// There are two kinds of dedicated refinement threads, a single primary
+// thread and some number of secondary threads.  When active, all refinement
+// threads take buffers of dirty cards from the dirty card queue and process
+// them.  Between buffers they query this owning object to find out whether
+// they should continue running, deactivating themselves if not.
+//
+// The primary thread drives the control system that determines how many
+// refinement threads should be active.  If inactive, it wakes up periodically
+// to recalculate the number of active threads needed, and activates
+// additional threads as necessary.  While active it also periodically
+// recalculates the number wanted and activates more threads if needed.  It
+// also reduces the number of wanted threads when the target has been reached,
+// triggering deactivations.
 class G1ConcurrentRefine : public CHeapObj<mtGC> {
+  G1Policy* _policy;
+  volatile uint _threads_wanted;
+  size_t _pending_cards_target;
+  Ticks _last_adjust;
+  Ticks _last_deactivate;
+  bool _needs_adjust;
+  G1ConcurrentRefineThreadsNeeded _threads_needed;
```

Comment:
`_threads_wanted` vs. `_threads_needed` makes for difficult reading in many places

Reply:
The intent is that "needed" refers to the number required to reach the goal,
while "wanted" is the number the controller wants active.  "wanted" may be
less than "needed" for a variety of reasons.

There were a couple of places where that intended nomenclature wasn't
followed, which I've fixed.

There is still G1ConcurrentRefine::_threads_wanted (the number of active
threads wanted by the controller) and G1ConcurrentRefine::_threads_needed (the
object that provides the calculation of the needed threads). From usage I
hope there isn't confusion, but I'm open to suggestions for better names.